### ISSUE TITLE
Bug: @typescript-eslint/no-unnecessary-condition false positives for getters

### ISSUE THREAD

**Issue by nbilyk** [typescript-eslint/typescript-eslint#8806]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.2&showAST=ts&fileType=.tsx&code=MYGwhgzhAEBiD29oG8BQrrQOYFMAu0ARmAE4AUAlAFxGIg5gB2KGm0J%2BAric2QLJg8ACwB0JJgBN4AW0rQAfNAAMIgKwVWAX1TbUweIwgEAZomgBeaIxwB3OIkroAlsehlT8EcRIUWmF24eXqS%2BaGzQ%2Bobw9CIg8FhkAOQAEgCSiRqY2ppAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class Foo {
  get bar(): boolean {
    return (Math.random() > 0.5)
  }
}

const foo = new Foo()

if (foo.bar) {
  if (foo.bar) { // ESLint Error
    console.log('HI')
  }
}
```
```


### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### Expected Result

Control flow analysis should understand that getters can change.

### Actual Result

Member type is incorrectly narrowed after first access.

**Comment by bradzacher** [comment]
This is working as expected.
The rule is powered by TS types and TS does not track impure getters with side-effects.
If you hover over the second member expression you'll see that the type is no `boolean` because TS has refined the type.

In general getters with side-effects are considered a large code-smell and an anti-pattern. Many static analysis and build tools are built to assume that property access is pure and consistent.

Instead of encoding such logic into a getter - you should use a method. All tools make no assumptions about method calls.

**Comment by nbilyk** [comment]
Ah, so it is. I didn't realize TS narrowed getters. Thanks for the fast response.

**Comment by nbilyk** [comment]
My example scenario was poor in that it had a side effect, but here's an example without one:

https://typescript-eslint.io/play/#ts=5.0.2&showAST=ts&fileType=.tsx&code=MYGwhgzhAEBiD29oG8BQrrQA4CcCWAbmAC4Cm0A%2BgEZg7QC80AZmCBKepgOanHQ04AFAEoAXP0QhSYAHYoMmaDl4BXHHOIALPBAB01WgoC%2BnaCqwATEqUFEQK0uKqTpM4fMXQtO-QIbQ7B2NUE1RgeBkIPiZEfxlSAHc4RBF0PCZoQRj4XQF3NExs3XMrMizWdmEFdMyivI9FcMj4KV0QeC5BAHIACQBJLqrMEyMgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

TypeScript will understand that after `foo.update(false)` foo.bar should be widened back to boolean.

It is an eslint bug.

**Comment by kirkwaiblinger** [comment]
@nbilyk When i hover over the second foo.bar, TS says: `(property) Foo.bar: true`.

**Comment by nbilyk** [comment]
Ah. Interesting. I'd expect TS control flow to widen the type but you're right, my mistake. 

Maybe I'll file a ticket with TS.
```
class Foo {
    bar = false
}

const foo = new Foo()

if (foo.bar) {
    const isTrue = foo.bar
    update(foo)
    const isAlsoTrue = foo.bar // unexpectedly type true
}

function update(foo: Foo) {
    foo.bar = false
}
```

**Comment by Josh-Cena** [comment]
I think this is microsoft/TypeScript#9998 (THE issue)

**Comment by bradzacher** [comment]
Flow treats all function calls as potential side-effects and invalidates all refinements.

TS does not - TS assumes the exact opposite.

Both approaches have drawbacks - flow prioritises correctness over devx, TS prirotises DevX over correctness.

The best solution would be to track side-effects but it's a really hard problem to solve.

Honestly I prefer flow's approach - it doesn't take long to get used to it and assign the refined value before a function call. But I understand many people hat writing code like that because they see it as code just to make the checker happy.

**Comment by nbilyk** [comment]
Yeah, agreed. It would make more sense to start assuming every call can mutate the value. More forward compatible as control flow improves. Kotlin does a very nice job with that, eventually adding contracts and the like.

Anyway, I'm not meaning to drive a TS discussion here as interesting as it is. Glad to see such an active community, especially on a holiday weekend :)
